gamemode/modules/base/sv_gamemode_functions.lua

```diff
@@ -312,40 +313,107 @@ for _, ply in pairs(player.GetHumans()) do
+    grid = {}
 
-hook.Add("PlayerDisconnected", "DarkRPCanHearVoice", function(ply)
-    DrpCanHear[ply] = nil
-    if ply:IsBot() then return end
-    for _, v in ipairs(player.GetHumans()) do
-        if not DrpCanHear[v] then continue end
-        DrpCanHear[v][ply] = nil
+    local plyPos = {}
+    local eyePos = {}
+
+    -- Get the grid position of every player O(N)
+    for _, ply in ipairs(players) do
+        local pos = ply:GetPos()
+        plyPos[ply] = pos
+        eyePos[ply] = ply:EyePos()
+        local x = floor(pos.x / gridSize)
+        local y = floor(pos.y / gridSize)
+
+        local row = grid[x] or {}
+            local cell = row[y] or {}
+                table.insert(cell, ply)
+            row[y] = cell
+        grid[x] = row
+
+        plyToGrid[1][ply] = x
+        plyToGrid[2][ply] = y
+
+        DrpCanHear[ply] = {} -- Initialize output variable
     end
-    timer.Remove(ply:UserID() .. "DarkRPCanHearPlayersVoice")
+
+    -- Check all neighbouring cells for every player.
+    -- We are only checking in 1 direction to avoid duplicate check of cells
+    for _, ply1 in ipairs(players) do
+        local gridX = plyToGrid[1][ply1]
+        local gridY = plyToGrid[2][ply1]
+        local ply1Pos = plyPos[ply1]
+        local ply1EyePos = eyePos[ply1]
+
+        for x = gridX, gridX + 1 do   -- Get every neighbouring cell (including the actual cell) (x axis)
+            local row = grid[x]
+            if not row then continue end
+
+            for y = gridY, gridY + 1 do   -- Get every neighbouring cell (including the actual cell) (y axis)
+                if x == gridX and y == gridY then continue end -- Exclude the actual cell as it needs a separate computation
+                local cell = row[y]
+                if not cell then continue end -- Check that values exist
+
+                -- Check if every player can talk
+                for _, ply2 in ipairs(cell) do
+                    local canTalk = not vrad or -- Voiceradius is off, everyone can hear everyone
+                        (ply1Pos:DistToSqr(plyPos[ply2]) < voiceDistance and -- voiceradius is on and the two are within hearing distance
+                            (not dynv or IsInRoom(ply1EyePos, eyePos[ply2], ply2))) -- Dynamic voice is on and players are in the same room
+
+                    DrpCanHear[ply1][ply2] = canTalk and (deadv or ply2:Alive())
+                    DrpCanHear[ply2][ply1] = canTalk and (deadv or ply1:Alive()) -- Take advantage of the symmetry
+                end
+            end
+        end
+    end
+
+    -- Doing a pass-through inside every cell to compute the interactions inside of the cells.
+    -- Each grid check is O(N(N+1)/2) where N is the number of players inside the cell.
+    for _, row in pairs(grid) do
+        for _, cell in pairs(row) do
+            local count = #cell
+            for i = 1, count do
+                for j = i + 1, count do
+                    local ply1, ply2 = cell[i], cell[j]
```

Comment:
As @FPtje said, you can move the ply1 down a scope. This will save a lot of indexing.